### PR TITLE
Hiding secrets in palette information - Part 1 (PNG)

### PR DESCRIPTION
**Pull request by jakubzehner**
- Added encoding text to png metadata
- Added decoding text from png metadata

### CODE REVIEW

File: code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+  public async encode(
+    inputPng: File,
+    key: string,
+    text: string,
+  ): Promise<File> {
+    const keyWithPrefix = this.createKey(key);
+    const inputBuffer = await inputPng.arrayBuffer().then((arrayBuffer) => {
+      return Buffer.from(arrayBuffer);
+    });
```
Comment (mlodybercik) [inline code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts:30]: Why not just this?

```suggestion
    const arrayBuffer = await inputPng.arrayBuffer();
    const inputBuffer = Buffer.from(arrayBuffer);
```

or even

```suggestion
    const inputBuffer = Buffer.from(await inputPng.arrayBuffer());
```

File: code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  public async decode(
+    inputPng: File,
+    key: string,
+  ): Promise<string | undefined> {
+    const keyWithPrefix = this.createKey(key);
+    const inputBuffer = await inputPng.arrayBuffer().then((arrayBuffer) => {
+      return Buffer.from(arrayBuffer);
+    });
```
Comment (mlodybercik) [inline code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts:60]:
```suggestion
    const arrayBuffer = await inputPng.arrayBuffer();
    const inputBuffer = Buffer.from(arrayBuffer);
```

```suggestion
    const inputBuffer = Buffer.from(await inputPng.arrayBuffer());
```

**Review by mlodybercik — APPROVED** [review]
It looks good. Few nitpicks as always.

**Comment by mlodybercik** [comment]
And also, I did not mention this in a review (because it does not have anything to do with that), but we should do #38. Image upload in PNG decoder could take `.jpeg` of `.webp` file. Which is obviously bad.

File: code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  private readonly CHUNK_TYPE = Buffer.from(this.CHUNK_TYPE_VALUE, "ascii");
+  private readonly CHUNK_TYPE_LENGTH = 4;
+  private readonly CHUNK_LENGTH_LENGTH = 4;
+  private readonly CHUNK_CRC_LENGTH = 4;
+  private readonly IEND_LENGTH = 12;
+  private readonly PNG_HEADER_LENGTH = 8;
+
+  private createKey = (key: string) => `${this.KEY_PREFIX}${key}`;
+
+  constructor() {}
```
Comment (tchojnacki) [inline code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts:20]: **question:** Doesn't JS define a zero-argument default constructor if there are no explicitly defined constructors?

File: code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
+import { Injectable } from "@angular/core";
+import { Buffer } from "buffer";
+import * as crc32 from "crc-32";
+
+@Injectable({
+  providedIn: "root",
+})
+export class PngMetaService {
```
Comment (tchojnacki) [inline code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts:8]: This class scares me on many levels. Good job.

File: code/projects/app/src/app/stages/e5/e5-decode/e5-decode.component.css [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  font-weight: bold;
+}
+
+section>textarea {
+  width: 100%;
+  background: var(--app-white);
+  border: 1px solid var(--app-black);
+}
+
+section>textarea:disabled {
```
[the remark this answers is not in the document]
Comment (mlodybercik) [reply code/projects/app/src/app/stages/e5/e5-decode/e5-decode.component.css:34]: Weird, locally it does add spaces. Maybe its the same issue we had with different versions of prettier being used by VSCode and installed in `node_modules`. 

When I run `prettier --check .` against this branch it complains only about CSS files. Maybe @jakubzehner has wrong formatter set for `.css` files? Could you check?

File: code/projects/app/src/app/stages/e5/e5-decode/e5-decode.component.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  protected readonly keyContent = signal("");
+  private readonly debouncedKey = debouncedSignal(this.keyContent, 300);
+
+  protected readonly image = signal<ImageData | undefined>(undefined);
+  protected readonly file = signal<File | undefined>(undefined);
+
+  protected onKeyChange(event: Event): void {
+    const key = (event.target as HTMLInputElement).value;
+    this.keyContent.set(key);
+  }
```
[the remark this answers is not in the document]
Comment (jakubzehner) [reply code/projects/app/src/app/stages/e5/e5-decode/e5-decode.component.ts:26]: yes

File: code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts (hunk 1 above)
Comment (jakubzehner) [reply code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts:30]: ok

Comment (mlodybercik) [reply code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts:30]: image

File: code/projects/app/src/app/stages/e5/e5-encode/e5-encode.component.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  protected readonly keyContent = signal("");
+  protected readonly textContent = signal("");
+  private readonly debouncedKey = debouncedSignal(this.keyContent, 300);
+  private readonly debouncedText = debouncedSignal(this.textContent, 300);
+
+  protected readonly image = signal<ImageData | undefined>(undefined);
+  protected readonly file = signal<File | undefined>(undefined);
+
+  protected readonly newFile = asyncComputed<File | undefined>(
```
[the remark this answers is not in the document]
Comment (jakubzehner) [reply code/projects/app/src/app/stages/e5/e5-encode/e5-encode.component.ts:27]: done

File: code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts (hunk 4 above)
Comment (jakubzehner) [reply code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts:8]: dont care

File: code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts (hunk 3 above)
Comment (jakubzehner) [reply code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts:20]: yep, mb, removed

File: code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts (hunk 4 above)
Comment (mlodybercik) [reply code/projects/app/src/app/stages/e5/e5-services/png-meta.service.ts:8]: dont-care-didnt-ask

**Pull request merged**